File: plugins/operators/schema_generator.py
"""
Operator to generate Zod and database schemas from TypeSpec files
"""

from typing import Any, Dict, Optional, Sequence

from airflow.models import BaseOperator
from airflow.utils.decorators import apply_defaults

from plugins.hooks.supabase_hook import SupabaseHook
from plugins.utils.node_executor import NodeExecutor


class SchemaGeneratorOperator(BaseOperator):
    """
    Operator that generates Zod and database schemas from TypeSpec files
    
    This operator takes a TypeSpec artifact, processes it using a Node.js script,
    and creates both Zod schema and database schema artifacts.
    """
    
    @apply_defaults
    def __init__(
        self,
        artifact_id: int,
        supabase_conn_id: str = 'supabase_default',
        *args, **kwargs
    ) -> None:
        """
        Initialize the schema generator operator
        
        Args:
            artifact_id: The ID of the TypeSpec artifact to process
            supabase_conn_id: The Supabase connection ID to use
        """
        super().__init__(*args, **kwargs)
        self.artifact_id = artifact_id
        self.supabase_conn_id = supabase_conn_id
    
    def execute(self, context: Dict[str, Any]) -> Dict[str, int]:
        """
        Execute the schema generation operation
        
        Args:
            context: The Airflow task context
            
        Returns:
            A dictionary with IDs of the created schema artifacts
        """
        # Get the TypeSpec artifact
        hook = SupabaseHook(self.supabase_conn_id)
        typespec_artifact = hook.get_artifact(self.artifact_id)
        
        if not typespec_artifact:
            raise ValueError(f"Artifact with ID {self.artifact_id} not found")
        
        if typespec_artifact['type'] != 'TYPESPEC':
            raise ValueError(f"Artifact with ID {self.artifact_id} is not a TypeSpec")
        
        # Process the TypeSpec to generate Zod and database schemas
        self.log.info(f"Generating schemas from TypeSpec: {typespec_artifact['name']}")
        
        # Execute the schema generator
        result = NodeExecutor.execute_processor('schema_generator', {
            'typespec': typespec_artifact['content']
        })
        
        # Create a Zod schema artifact
        zod_schema_artifact = {
            'name': f"Zod Schema for {typespec_artifact['name']}",
            'type': 'ZOD_SCHEMA',
            'content': result['zod_schema'],
            'createdBy': context['task_instance'].task_id,
            'parentId': typespec_artifact['id'],
            'status': 'PENDING_APPROVAL'
        }
        
        # Save the Zod schema artifact
        created_zod_artifact = hook.create_artifact(zod_schema_artifact)
        
        self.log.info(f"Created Zod schema artifact with ID: {created_zod_artifact['id']}")
        
        # Create a database schema artifact
        db_schema_artifact = {
            'name': f"Database Schema for {typespec_artifact['name']}",
            'type': 'DB_SCHEMA',
            'content': result['db_schema'],
            'createdBy': context['task_instance'].task_id,
            'parentId': typespec_artifact['id'],
            'status': 'PENDING_APPROVAL'
        }
        
        # Save the database schema artifact
        created_db_artifact = hook.create_artifact(db_schema_artifact)
        
        self.log.info(f"Created database schema artifact with ID: {created_db_artifact['id']}")
        
        # Get the pipeline stages for approvals
        zod_stage = hook.get_pipeline_stage_by_order(4)  # Zod Schema is stage 4
        db_stage = hook.get_pipeline_stage_by_order(5)   # DB Schema is stage 5
        
        # Create approval tasks for Zod schema
        zod_approval1 = {
            'artifactId': created_zod_artifact['id'],
            'stakeholderId': 2,  # This should be dynamically determined
            'status': 'PENDING'
        }
        
        zod_approval2 = {
            'artifactId': created_zod_artifact['id'],
            'stakeholderId': 4,  # This should be dynamically determined
            'status': 'PENDING'
        }
        
        hook.create_approval(zod_approval1)
        hook.create_approval(zod_approval2)
        
        # Create approval tasks for database schema
        db_approval1 = {
            'artifactId': created_db_artifact['id'],
            'stakeholderId': 3,  # This should be dynamically determined
            'status': 'PENDING'
        }
        
        db_approval2 = {
            'artifactId': created_db_artifact['id'],
            'stakeholderId': 5,  # This should be dynamically determined
            'status': 'PENDING'
        }
        
        hook.create_approval(db_approval1)
        hook.create_approval(db_approval2)
        
        return {
            'zod_schema_id': created_zod_artifact['id'],
            'db_schema_id': created_db_artifact['id']
        }

File: plugins/operators/schema_generator.js
/**
 * Airflow operator for generating Zod and Drizzle schemas from TypeSpec
 */

const { typeSpecToZod, typeSpecToDrizzle } = require('../../helpers/type_utils');
const SupabaseHook = require('../hooks/supabase_hook');

class SchemaGeneratorOperator {
  constructor(taskId, pipelineRunId, artifactId, stageId) {
    this.taskId = taskId;
    this.pipelineRunId = pipelineRunId;
    this.artifactId = artifactId;
    this.stageId = stageId;
    this.supabaseHook = new SupabaseHook();
  }

  /**
   * Execute the schema generation task
   * @returns {Promise<Object>} - The execution result
   */
  async execute() {
    try {
      console.log(`Executing schema generator task: ${this.taskId}`);
      
      // Create stage execution record
      const stageExecution = await this.supabaseHook.createStageExecution({
        pipelineRunId: this.pipelineRunId,
        stageId: this.stageId,
        artifactId: this.artifactId,
        status: 'running'
      });
      
      // Get the TypeSpec artifact
      const typeSpecArtifact = await this.supabaseHook.getArtifact(this.artifactId);
      if (!typeSpecArtifact) {
        throw new Error(`TypeSpec artifact with ID ${this.artifactId} not found`);
      }
      
      console.log(`Generating schemas from TypeSpec: ${typeSpecArtifact.title}`);
      
      // Extract TypeSpec content
      const { typeSpecText } = typeSpecArtifact.content;
      
      // Generate Zod schemas
      const zodSchemas = this.generateZodSchemas(typeSpecText);
      
      // Generate Drizzle schemas
      const drizzleSchemas = this.generateDrizzleSchemas(typeSpecText);
      
      // Create new artifact for schemas
      const schemaArtifact = await this.supabaseHook.createArtifact({
        title: `${typeSpecArtifact.title} - Schemas`,
        description: `Zod and Drizzle schemas generated from ${typeSpecArtifact.title}`,
        type: 'Schema',
        content: {
          zodSchemas,
          drizzleSchemas,
          metadata: {
            generatedAt: new Date().toISOString(),
            version: '1.0.0',
            source: typeSpecArtifact.id
          }
        },
        version: '1.0.0',
        parentId: typeSpecArtifact.id,
        status: 'draft'
      });
      
      // Update stage execution to completed
      await this.supabaseHook.updateStageExecution(stageExecution.id, {
        status: 'completed',
        endTime: new Date(),
        artifactId: schemaArtifact.id,
        logs: JSON.stringify({
          message: 'Successfully generated schemas',
          zodSchemasCount: Object.keys(zodSchemas).length,
          drizzleSchemasCount: Object.keys(drizzleSchemas).length
        })
      });
      
      return {
        status: 'success',
        artifactId: schemaArtifact.id,
        stageExecutionId: stageExecution.id
      };
    } catch (error) {
      console.error('Error in schema generator:', error);
      
      // Update stage execution to failed
      if (this.stageExecution) {
        await this.supabaseHook.updateStageExecution(this.stageExecution.id, {
          status: 'failed',
          endTime: new Date(),
          logs: JSON.stringify({
            error: error.message,
            stack: error.stack
          })
        });
      }
      
      throw error;
    }
  }

  /**
   * Generate Zod schemas from TypeSpec
   * @param {string} typeSpecText - TypeSpec content
   * @returns {Object} - Zod schemas for each model
   */
  generateZodSchemas(typeSpecText) {
    const zodSchemas = {};
    
    // Extract models from TypeSpec
    const modelRegex = /model\s+(\w+)\s+{([^}]+)}/g;
    let modelMatch;
    
    while ((modelMatch = modelRegex.exec(typeSpecText)) !== null) {
      const modelName = modelMatch[1];
      const modelDefinition = modelMatch[0];
      
      // Convert model to Zod schema
      const zodSchema = typeSpecToZod(modelDefinition);
      
      // Store schema
      zodSchemas[modelName] = zodSchema;
    }
    
    return zodSchemas;
  }

  /**
   * Generate Drizzle schemas from TypeSpec
   * @param {string} typeSpecText - TypeSpec content
   * @returns {Object} - Drizzle schemas for each model
   */
  generateDrizzleSchemas(typeSpecText) {
    const drizzleSchemas = {};
    
    // Extract models from TypeSpec
    const modelRegex = /model\s+(\w+)\s+{([^}]+)}/g;
    let modelMatch;
    
    while ((modelMatch = modelRegex.exec(typeSpecText)) !== null) {
      const modelName = modelMatch[1];
      const modelDefinition = modelMatch[0];
      
      // Convert model to Drizzle schema
      const drizzleSchema = typeSpecToDrizzle(modelDefinition);
      
      // Store schema
      drizzleSchemas[modelName] = drizzleSchema;
    }
    
    return drizzleSchemas;
  }
}

module.exports = SchemaGeneratorOperator;
